### PR TITLE
don't rerun prefixedChildPaths stream in aggregate

### PR DESCRIPTION
**Pull request by alissapajer**
[ch12096]

### CODE REVIEW

File: impl/src/main/scala/quasar/impl/datasource/AggregatingDatasource.scala [hunk 1]
```diff
@@ -66,18 +66,33 @@ final class AggregatingDatasource[F[_]: MonadResourceErr: Sync, Q, R] private(
...
 
   def prefixedChildPaths(prefixPath: ResourcePath)
-      : Resource[F, Option[Stream[F, (ResourceName, ResourcePathType)]]] =
+      : Resource[F, Option[Stream[F, (ResourceName, ResourcePathType)]]] = {
+
+    def loop[A](s: Stream[F, A]): Pull[F, A, Unit] =
+      s.pull.uncons flatMap {
+        case Some((hd, tl)) => Pull.output(hd) >> loop[A](tl)
+        case None => Pull.done
+      }
```
Comment (alissapajer) [inline impl/src/main/scala/quasar/impl/datasource/AggregatingDatasource.scala:75]: is there a shorthand/utility that solves this problem?

**Comment by alissapajer** [comment]
released as `188.0.1-5d279eb`

File: impl/src/main/scala/quasar/impl/datasource/AggregatingDatasource.scala (hunk 1 above)
Comment (wemrysi) [reply impl/src/main/scala/quasar/impl/datasource/AggregatingDatasource.scala:75]: Indeed, `s.pull.echo`

**Review by wemrysi — APPROVED** [review]

File: impl/src/main/scala/quasar/impl/datasource/AggregatingDatasource.scala [hunk 2]
```diff
@@ -61,46 +61,53 @@ final class AggregatingDatasource[F[_]: MonadResourceErr: Sync, Q, R] private(
         })
     }
 
-  def pathIsResource(path: ResourcePath): Resource[F, Boolean] =
-    underlying.pathIsResource(path)
-      .ifM(true.pure[Resource[F, ?]], aggPathExists(path))
+  def pathIsResource(path: ResourcePath): Resource[F, Boolean] = {
+    // `p / **` exists iff `p` is prefix/prefixresource
+    def aggPathExists: Resource[F, Boolean] = {
```
Comment (alissapajer) [inline impl/src/main/scala/quasar/impl/datasource/AggregatingDatasource.scala:66]: 👍 to moving this inside `pathIsResource` so it won't be used accidentally in the future

Comment (alissapajer) [reply impl/src/main/scala/quasar/impl/datasource/AggregatingDatasource.scala:66]: or really I meant the line below, `isNonEmpty` - and I had the thought to rename that as well :)

File: impl/src/main/scala/quasar/impl/datasource/AggregatingDatasource.scala [hunk 3]
```diff
@@ -61,46 +61,53 @@ final class AggregatingDatasource[F[_]: MonadResourceErr: Sync, Q, R] private(
...
-          Stream.empty.covary[F].covaryOutput[(ResourceName, ResourcePathType)].some.pure[Resource[F, ?]],
-          none.pure[Resource[F, ?]])
-    }
+    if (aggPath(prefixPath).isDefined)
+      pathIsResource(prefixPath).ifM(
+        ifTrue = Stream.empty.covary[F].covaryOutput[(ResourceName, ResourcePathType)].some.pure[Resource[F, ?]],
+        ifFalse = none.pure[Resource[F, ?]])
+    else
+      underlying.prefixedChildPaths(prefixPath).map(_.map(
+        _.pull.peek1.flatMap {
```
Comment (alissapajer) [inline impl/src/main/scala/quasar/impl/datasource/AggregatingDatasource.scala:94]: `peek1` is nicer since we don't have to resequence the head

**Pull request merged**